File: src/ai-node/bootstrap/bootstrapClient.js
import { parseBootstrapPayload, validateBootstrapPayload } from "./bootstrapParser.js";
import { NODE_LIFECYCLE_STATES } from "../lifecycle/nodeLifecycle.js";

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function assertExactTopic(topic) {
  if (typeof topic !== "string" || topic.length === 0) {
    throw new Error("bootstrap topic is required");
  }
  if (topic.includes("#") || topic.includes("+")) {
    throw new Error("wildcard bootstrap topic is not allowed");
  }
}

export function createBootstrapClient(options) {
  if (!options || typeof options !== "object") {
    throw new Error("bootstrap client options are required");
  }

  const lifecycle = options.lifecycle;
  if (!lifecycle || typeof lifecycle.transitionTo !== "function") {
    throw new Error("bootstrap client requires lifecycle controller");
  }

  const logger = options.logger ?? console;
  const mqttAdapter = options.mqttAdapter;
  if (!mqttAdapter || typeof mqttAdapter.connect !== "function") {
    throw new Error("bootstrap client requires mqttAdapter.connect");
  }

  const retryConfig = {
    maxAttempts: options.maxAttempts ?? 5,
    baseDelayMs: options.baseDelayMs ?? 500,
    maxDelayMs: options.maxDelayMs ?? 5_000,
  };

  let client = null;
  let isRunning = false;

  async function connect(config, callbacks = {}) {
    assertExactTopic(config.topic);
    isRunning = true;
    let attempt = 0;

    while (isRunning && attempt < retryConfig.maxAttempts) {
      attempt += 1;
      lifecycle.transitionTo(NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTING, { attempt });

      try {
        client = await mqttAdapter.connect({
          host: config.bootstrapHost,
          port: config.port,
          username: undefined,
          password: undefined,
          clientId: config.nodeName,
          clean: true,
        });

        await client.subscribe(config.topic);
        lifecycle.transitionTo(NODE_LIFECYCLE_STATES.BOOTSTRAP_CONNECTED, { attempt });
        logger.info?.("[bootstrap-connected]", {
          host: config.bootstrapHost,
          port: config.port,
          topic: config.topic,
        });

        client.onMessage((topic, rawPayload) => {
          if (topic !== config.topic) {
            return;
          }

          const parsed = parseBootstrapPayload(rawPayload);
          if (!parsed.ok) {
            logger.warn?.("[bootstrap-payload-ignored]", { reason: parsed.error });
            return;
          }

          const validated = validateBootstrapPayload(parsed.value, {
            expectedTopic: config.topic,
            supportedVersions: callbacks.supportedBootstrapVersions ?? [1],
          });
          if (!validated.ok) {
            logger.warn?.("[bootstrap-payload-ignored]", { reason: validated.error });
            return;
          }

          lifecycle.transitionTo(NODE_LIFECYCLE_STATES.CORE_DISCOVERED);
          callbacks.onCoreDiscovered?.(validated.value);
        });

        return {
          close: async () => {
            isRunning = false;
            await client?.close?.();
            client = null;
          },
        };
      } catch (error) {
        logger.warn?.("[bootstrap-connect-failed]", {
          attempt,
          maxAttempts: retryConfig.maxAttempts,
          message: error?.message ?? "unknown error",
        });

        if (attempt >= retryConfig.maxAttempts) {
          lifecycle.transitionTo(NODE_LIFECYCLE_STATES.DEGRADED, { stage: "bootstrap_connect" });
          throw new Error("bootstrap connection failed after bounded retries");
        }

        const nextDelayMs = Math.min(
          retryConfig.baseDelayMs * Math.pow(2, attempt - 1),
          retryConfig.maxDelayMs,
        );
        await delay(nextDelayMs);
      }
    }

    throw new Error("bootstrap client stopped");
  }

  return {
    connect,
    stop: async () => {
      isRunning = false;
      await client?.close?.();
      client = null;
    },
  };
}

File: src/ai_node/bootstrap/bootstrap_client.py
import asyncio
from typing import Callable, Optional

from ai_node.bootstrap.bootstrap_parser import parse_bootstrap_payload, validate_bootstrap_payload
from ai_node.diagnostics.onboarding_logger import OnboardingDiagnosticsLogger
from ai_node.lifecycle.node_lifecycle import NodeLifecycle, NodeLifecycleState


def _assert_exact_topic(topic: str) -> None:
    if not isinstance(topic, str) or not topic:
        raise ValueError("bootstrap topic is required")
    if "#" in topic or "+" in topic:
        raise ValueError("wildcard bootstrap topic is not allowed")


class BootstrapClient:
    def __init__(
        self,
        *,
        lifecycle: NodeLifecycle,
        mqtt_adapter,
        logger,
        max_attempts: int = 5,
        base_delay_seconds: float = 0.5,
        max_delay_seconds: float = 5.0,
    ) -> None:
        if lifecycle is None:
            raise ValueError("bootstrap client requires lifecycle controller")
        if mqtt_adapter is None or not hasattr(mqtt_adapter, "connect"):
            raise ValueError("bootstrap client requires mqtt_adapter.connect")

        self._lifecycle = lifecycle
        self._mqtt_adapter = mqtt_adapter
        self._logger = logger
        self._diag = OnboardingDiagnosticsLogger(logger)
        self._max_attempts = max_attempts
        self._base_delay_seconds = base_delay_seconds
        self._max_delay_seconds = max_delay_seconds
        self._client = None
        self._running = False

    async def connect(
        self,
        config,
        *,
        on_core_discovered: Optional[Callable[[dict], None]] = None,
        supported_bootstrap_versions=(1,),
    ):
        _assert_exact_topic(config.topic)
        self._running = True
        attempt = 0

        while self._running and attempt < self._max_attempts:
            attempt += 1
            self._lifecycle.transition_to(NodeLifecycleState.BOOTSTRAP_CONNECTING, {"attempt": attempt})
            try:
                self._client = await self._mqtt_adapter.connect(
                    {
                        "host": config.bootstrap_host,
                        "port": config.port,
                        "username": None,
                        "password": None,
                        "client_id": config.node_name,
                        "clean": True,
                    }
                )
                await self._client.subscribe(config.topic)
                self._lifecycle.transition_to(NodeLifecycleState.BOOTSTRAP_CONNECTED, {"attempt": attempt})
                if hasattr(self._logger, "info"):
                    self._logger.info(
                        "[bootstrap-connected] %s",
                        {"host": config.bootstrap_host, "port": config.port, "topic": config.topic},
                    )
                self._diag.bootstrap_connect(
                    {"host": config.bootstrap_host, "port": config.port, "topic": config.topic}
                )

                async def _handle_message(topic: str, raw_payload: object) -> None:
                    if topic != config.topic:
                        return

                    parsed_ok, parsed_value = parse_bootstrap_payload(raw_payload)
                    if not parsed_ok:
                        self._diag.payload_validation({"result": "ignored", "reason": parsed_value})
                        if hasattr(self._logger, "warning"):
                            self._logger.warning("[bootstrap-payload-ignored] %s", {"reason": parsed_value})
                        return

                    valid_ok, valid_value = validate_bootstrap_payload(
                        parsed_value,
                        expected_topic=config.topic,
                        supported_versions=supported_bootstrap_versions,
                    )
                    if not valid_ok:
                        self._diag.payload_validation({"result": "ignored", "reason": valid_value})
                        if hasattr(self._logger, "warning"):
                            self._logger.warning("[bootstrap-payload-ignored] %s", {"reason": valid_value})
                        return

                    self._diag.payload_validation({"result": "accepted", "core_id": valid_value.get("core_id")})
                    self._lifecycle.transition_to(NodeLifecycleState.CORE_DISCOVERED)
                    if callable(on_core_discovered):
                        on_core_discovered(valid_value)

                self._client.on_message(_handle_message)
                return self
            except Exception as exc:
                if hasattr(self._logger, "warning"):
                    self._logger.warning(
                        "[bootstrap-connect-failed] %s",
                        {
                            "attempt": attempt,
                            "max_attempts": self._max_attempts,
                            "message": str(exc),
                        },
                    )
                if attempt >= self._max_attempts:
                    self._lifecycle.transition_to(
                        NodeLifecycleState.DEGRADED,
                        {"stage": "bootstrap_connect"},
                    )
                    raise RuntimeError("bootstrap connection failed after bounded retries") from exc

                delay_seconds = min(
                    self._base_delay_seconds * (2 ** (attempt - 1)),
                    self._max_delay_seconds,
                )
                await asyncio.sleep(delay_seconds)

        raise RuntimeError("bootstrap client stopped")

    async def stop(self) -> None:
        self._running = False
        if self._client is not None and hasattr(self._client, "close"):
            await self._client.close()
            self._diag.bootstrap_disconnect({"reason": "stop_called"})
            self._client = None
